internal/app/machined/pkg/controllers/config/node_labels.go

```diff
@@ -0,0 +1,334 @@
+
+		k8sLabelValue, ok := k8sLabels[stateLabelSpec.Key]
+		if ok {
+			delete(k8sLabels, stateLabelSpec.Key)
+
+			if k8sLabelValue != stateLabelSpec.Value {
+				if err := updateLabel(ctx, r, stateLabel, k8sLabelValue, false); err != nil {
+					return fmt.Errorf("error updating node label resource: %w", err)
+				}
+			}
+		}
+	}
+
+	return addLabelsToState(ctx, r, k8sLabels, mcNodeLabels)
+}
+
+func addLabelsToState(ctx context.Context, r controller.Runtime, k8sLabels, mcNodeLabels map[string]string) error {
+	for key, value := range k8sLabels {
+		createdByTalos := false
+
+		mcValue, ok := mcNodeLabels[key]
+		if ok && mcValue == value {
+			createdByTalos = true
+		}
+
+		labelRes := config.NewPopulatedNodeLabel(key, key, value, createdByTalos)
+
+		if err := r.Create(ctx, labelRes); err != nil {
+			return fmt.Errorf("failed to create node label resource: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func copyMap(src map[string]string) map[string]string {
+	dst := map[string]string{}
+
+	for key, value := range src {
+		dst[key] = value
+	}
+
+	return dst
+}
+
+func (ctrl *NodeLabelsController) computeChangesToMake(stateNodeLabels *resource.List, mcNodeLabels talosconfig.NodeLabels) (map[string]struct{}, map[string]string, map[string]string) {
+	labelIDsToDelete := map[resource.ID]struct{}{}
+
+	mcNodeLabelsMap := map[string]string{}
+	labelsToAdd := map[string]string{}
+	labelsToChange := map[string]string{}
+
+	for key, value := range mcNodeLabels {
+		mcNodeLabelsMap[key] = value
+		labelsToAdd[key] = value
+	}
+
+	for _, stateLabel := range stateNodeLabels.Items {
+		stateLabelSpec := stateLabel.(*config.NodeLabel).TypedSpec()
+
+		mcLabelValue, ok := mcNodeLabelsMap[stateLabelSpec.Key]
+		if ok {
+			delete(labelsToAdd, stateLabelSpec.Key)
+
+			if mcLabelValue != stateLabelSpec.Value {
+				labelsToChange[stateLabelSpec.Key] = mcLabelValue
+			}
+		} else if stateLabelSpec.CreatedByTalos {
+			labelIDsToDelete[stateLabel.Metadata().ID()] = struct{}{}
+		}
+	}
+
+	return labelIDsToDelete, labelsToAdd, labelsToChange
+}
+
+func (ctrl *NodeLabelsController) reconcileNodeLabels(
+	ctx context.Context,
+	r controller.Runtime,
+	nodeName string,
+	mcNodeLabels talosconfig.NodeLabels,
+	k8sClient NodeLabelsK8sClientIface,
+	logger *zap.Logger,
+) error {
+	stateNodeLabels, err := getStateNodeLabels(ctx, r)
+	if err != nil {
+		return err
+	}
+
+	labelIDsToDelete, labelsToAdd, labelsToChange := ctrl.computeChangesToMake(stateNodeLabels, mcNodeLabels)
+
+	// We are making a compromise here.
+	// Since we don't have transactions in COSI, we can't make sure that both the labels on the kubernetes Node
+	// aswell as in the COSI state are the same in case writing one of them fails.
+	// To make sure labels are always applied to kubernetes we write those first and update the COSI state second.
+	// In case of an error while writing to COSI state, the labels will be picked up on the next reconcile loop,
+	// but will be marked as not created by Talos.
```

Comment:
Not suggesting to change it but, couldn't we instead use a well-known prefix like `talos.dev/` so all labels managed by Talos would require to be in the form `talos.dev/some-label: value`? It would simplify detecting the owner, we probably wouldn't even need a dedicated resource?

If we don't want the prefix, then I think we could do something simple as, the labels on machineconfig would always win over (overwrite) the other sources.

Reply:
I think we can't limit our users to use labels of some form, they might need to set labels the way they want it